openlabcmd/openlabcmd/zk.py

```diff
@@ -0,0 +1,169 @@
+            self._last_retry_log = now
+
+    @property
+    def connected(self):
+        if self.client is None:
+            return False
+        return self.client.state == KazooState.CONNECTED
+
+    @property
+    def suspended(self):
+        if self.client is None:
+            return True
+        return self.client.state == KazooState.SUSPENDED
+
+    @property
+    def lost(self):
+        if self.client is None:
+            return True
+        return self.client.state == KazooState.LOST
+
+    def connect(self, hosts=None, read_only=False):
+        if not hosts:
+            if not self.config:
+                raise Exception('Either config object or hosts string should '
+                                'be provided.')
+            else:
+                try:
+                    hosts = self.config.get('ha', 'zookeeper_hosts')
+                except configparser.NoOptionError:
+                    raise Exception("The config doesn't [ha]zookeeper_hosts "
+                                    "option")
+        if self.client is None:
+            self.client = KazooClient(hosts=hosts, read_only=read_only)
+            self.client.add_listener(self._connection_listener)
+            # Manually retry initial connection attempt
+            while True:
+                try:
+                    self.client.start(1)
+                    break
+                except KazooTimeoutError:
+                    self.logConnectionRetryEvent()
+
+    def disconnect(self):
+        if self.client is not None and self.client.connected:
+            self.client.stop()
+            self.client.close()
+            self.client = None
+
+    def _client_check_wrapper(func):
+        def wrapper(self, *args, **kwargs):
+            if not self.client:
+                raise Exception("should call connect function first to "
+                                "initialise zookeeper client")
+            return func(self, *args, **kwargs)
+        return wrapper
+
+    @_client_check_wrapper
+    def list_nodes(self):
+        path = '/ha'
+        try:
+            nodes = []
+            for node in self.client.get_children(path):
+                node_info = self.client.get('/ha/%s' % node)
+                node_info = self._bytesToDict(node_info[0])
+                node_info['name'] = node
+                nodes.append(node_info)
+        except kze.NoNodeError:
+            return []
+        return sorted(nodes)
+
+    @_client_check_wrapper
+    def get_node(self, node_name):
+        try:
+            node_info = self.client.get('/ha/%s' % node_name)
+            node_info = self._bytesToDict(node_info[0])
+            node_info['name'] = node_name
+            return node_info
+        except kze.NoNodeError:
+            return {}
+
+    @_client_check_wrapper
+    def create_node(self, node_name, node_info):
+        path = '/ha/%s' % node_name
+        node_info['alarmed'] = False
+        node_info['status'] = 'initializing'
```

Comment:
How to update node status from `initializing` to `active`?

Reply:
Keepalived will do that.

Reply:
OK, thank you.